### PR TITLE
chore: sanitize the .NET codebase (a little)

### PR DESCRIPTION
**Pull request by RomainMuller**
Configured a bunch of Roslyn analyzers to ensure code quality, and made the necessary corrections:

- Correct usage of `ArgumentException`
- Use `Array.Empty<T>()` to spare allocations of empty objects
- Override equality on data types
- Specify `InvariantCulture` where needed (could be the source of weird bugs)
- Some other miscellaneous tweaks

Additionally, folded all the contents of `Amazon.JSII.JsonModel` under the `Amazon.JSII.Runtime`
assembly, and issued type frowards in `Amazon.JSII.JsonModel` to maintain backwards compatibility.

---

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license].

[Apache 2.0 license]: https://www.apache.org/licenses/LICENSE-2.0

### CODE REVIEW

File: packages/@jsii/dotnet-runtime/src/Amazon.JSII.Analyzers.UnitTests/Helpers/DiagnosticResult.cs [hunk 1]
```diff
@@ -54,10 +89,50 @@ public DiagnosticResultLocation[] Locations
...
+        }
+
+        public bool Equals(DiagnosticResult other)
+        {
+            return Id == other.Id
+                   && Severity == other.Severity
+                   && Message == other.Message
+                   && Path == other.Path
+                   && Line == other.Line
+                   && Column == other.Column;
```
Comment (njlynch) [inline packages/@jsii/dotnet-runtime/src/Amazon.JSII.Analyzers.UnitTests/Helpers/DiagnosticResult.cs:125]: This is making the assumption that two `DiagnosticResults` with the same first `DiagnosticResultLocation` are the same, regardless of the other `DiagnosticResultLocation` in `Locations`. I lack context to know if that's valid or not; I _suspect_ it's ok, but worth calling it out.

File: packages/@jsii/dotnet-runtime/src/Amazon.JSII.Analyzers/JsiiOptionalAnalyzer.cs [hunk 2]
```diff
@@ -15,8 +16,8 @@ public class JsiiOptionalAnalyzer : DiagnosticAnalyzer
         private const string Title = "A required property is missing or null";
         private const string MessageFormat = "The property is required and cannot be null";
         private const string MessageFormatWithPropertyName = "The property {0} is required and cannot be null";
```
Comment (njlynch) [inline packages/@jsii/dotnet-runtime/src/Amazon.JSII.Analyzers/JsiiOptionalAnalyzer.cs:18]: You added periods to the two below messages; any reason not to add them to these?

File: packages/@jsii/dotnet-runtime/src/Amazon.JSII.Analyzers/JsiiOptionalAnalyzer.cs [hunk 3]
```diff
@@ -15,8 +16,8 @@ public class JsiiOptionalAnalyzer : DiagnosticAnalyzer
...
-        private const string Description = "The property is required and cannot be null";
-        private const string DescriptionWitPropertyName = "The property {0} is required and cannot be null";
+        private const string Description = "The property is required and cannot be null.";
+        private const string DescriptionWitPropertyName = "The property {0} is required and cannot be null.";
```
Comment (njlynch) [inline packages/@jsii/dotnet-runtime/src/Amazon.JSII.Analyzers/JsiiOptionalAnalyzer.cs:20]:
```suggestion
        private const string DescriptionWithPropertyName = "The property {0} is required and cannot be null.";
```

File: packages/@jsii/dotnet-runtime/src/Amazon.JSII.Analyzers/JsiiOptionalAnalyzer.cs (hunk 2 above)
Comment (RomainMuller) [reply packages/@jsii/dotnet-runtime/src/Amazon.JSII.Analyzers/JsiiOptionalAnalyzer.cs:18]: Basically it's the following rules from the Roslyn analyzer for Roslyn analyzers (so meta):
- [RS1033](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.CodeAnalysis.Analyzers/Microsoft.CodeAnalysis.Analyzers.md#rs1033-define-diagnostic-description-correctly)
  > The diagnostic description should be one or multiple sentences ending with a punctuation sign
- [RS1032](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.CodeAnalysis.Analyzers/Microsoft.CodeAnalysis.Analyzers.md#rs1032-define-diagnostic-message-correctly)
  > The diagnostic message should not contain any line return character and should either be a single sentence without a trailing period or a multi-sentences with a trailing period

**Pull request merged**